### PR TITLE
Common Error Design

### PR DESCRIPTION
**Pull request by symorton**
**Problem:**
We need a way to standardize errors within the library and also between iOS and Android library.

**Solution:**
This PR proposes a factory-like solution to make all externally thrown errors conform to the same base class, and be created by a `VerifiedIdErrors` enum. There is also added logic to make logging errors more friendly and error codes to help with debugging. The description for the errors will now be Stringified JSON with all properties included.

Here is a basic example:
```JSON
{
"code": "requirement_not_met",
"message": "Pin has not been set",
}
```

Also included is a new object called `VerifiedIdResult` that will be the return value for all external interfaces. It will return a `VerifiedIdError` if there is an error instead of just a more generic `Error`.

**Validation:**
The `PinRequirement` and one external method on `VerifiedIdClient` class was modified to test out the design.

**Type of change:**
- [X] Feature work
- [ ] Bug fix
- [ ] Documentation
- [X] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

**Risk**:
- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [X] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

**Work Item links**:
Please include here links for this work item, or deferred work, or related work. E.g. if the refactoring is too big to fit in this PR, or the localized strings need to be updated later, please link the TODO work items here.

**Documentation Links**:
Please include here links to any related background documentation for this PR.

### CODE REVIEW

File: WalletLibrary/WalletLibrary/VerifiedIdClient.swift [hunk 1]
```diff
@@ -23,25 +23,28 @@ public class VerifiedIdClient {
...
     
     /// Encode a VerifiedId into Data.
-    public func encode(verifiedId: VerifiedId) -> Result<Data, Error> {
+    public func encode(verifiedId: VerifiedId) -> VerifiedIdResult<Data> {
         do {
             let encodedVerifiedId = try configuration.verifiedIdEncoder.encode(verifiedId: verifiedId)
             return Result.success(encodedVerifiedId)
         } catch {
-            return Result.failure(error)
+            return Result.failure(VerifiedIdErrors.UnspecifiedError(error: error).error)
```
Comment (logangirvin) [inline WalletLibrary/WalletLibrary/VerifiedIdClient.swift]: I'm questioning if we could we put a convivence init on `VerifiedIdResult` that calls the `VerifiedIdErrors.error` for us?  
 
Would this even be preferred for clarity?
```
return VerifiedIdResult.failure(VerifiedIdErrors.UnspecifiedError(error: error))
```

**Review by logangirvin — APPROVED** [review]
Overall looks good to me but I'd wait to see what Daniel or Kashif have to say.

File: WalletLibrary/WalletLibrary/VerifiedIdClient.swift (hunk 1 above)
Comment (symorton) [reply WalletLibrary/WalletLibrary/VerifiedIdClient.swift]: yes wow I literally was just thinking the same thing. Playing around with it now.

Comment (symorton) [reply WalletLibrary/WalletLibrary/VerifiedIdClient.swift]: Okay @logangirvin I added a couple of helper functions for this. What do you think?

File: WalletLibrary/WalletLibrary/Errors/VerifiedIdErrors.swift [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    
+    /// Common Error Codes.
+    struct ErrorCode {
+        static let NetworkingError = "networking_error"
+        static let RequirementNotMet = "requirement_not_met"
+        static let UnspecifiedError = "unspecified_error"
+    }
+    
+    /// Common Errors in Alphabetical Order.
+    case NetworkingError(message: String, correlationId: String, statusCode: String? = nil, innerError: Error? = nil)
```
Comment (ksaaf) [inline WalletLibrary/WalletLibrary/Errors/VerifiedIdErrors.swift:18]: Why not have optional 'correlationId' for other error types in addition to NetworkingError?

Comment (symorton) [reply WalletLibrary/WalletLibrary/Errors/VerifiedIdErrors.swift:18]: good point. updated!

**Pull request merged**